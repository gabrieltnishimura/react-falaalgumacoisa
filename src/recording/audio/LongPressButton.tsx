import React, { useEffect, useRef, useState } from 'react';
import { fromEvent, race } from 'rxjs';
import { debounceTime, delay } from 'rxjs/operators';
import styles from './LongPressButton.module.css';

type PressedType = React.MouseEvent<any> | React.TouchEvent<any>;
interface LongPressButtonInput {
  pressed: (e: PressedType) => void;
  unpressed: (e: PressedType) => void;
}

function LongPressButton(props: LongPressButtonInput) {
  const [recording, setRecording] = useState<boolean>(false);
  const imageRef = useRef<HTMLDivElement>(null);
  const componentPressed = (e: PressedType) => {
    setRecording(true);
    props.pressed(e);
  }

  const componentUnpressed = (e: PressedType) => {
    setRecording(false);
    props.unpressed(e);
  }

  useEffect(() => {
    if (!imageRef || !imageRef.current) {
      return;
    }

    const mouseDown = fromEvent<PressedType>(imageRef.current, 'mousedown', { passive: true });
    const touchStart = fromEvent<PressedType>(imageRef.current, 'touchstart', { passive: true });
    const pressedSubscription = race(mouseDown, touchStart)
      .pipe(debounceTime(1))
      .subscribe(componentPressed);

    const mouseUp = fromEvent<PressedType>(imageRef.current, 'mouseup', { passive: true });
    const touchEnd = fromEvent<PressedType>(imageRef.current, 'touchend', { passive: true });
    const unpressedSubscription = race(mouseUp, touchEnd)
      .pipe(
        debounceTime(1),
        delay(10), // what why
      )
      .subscribe(componentUnpressed);

    return () => {
      pressedSubscription.unsubscribe();
      unpressedSubscription.unsubscribe();
    };
  });

  const ignoreContextDropdown = (e: React.MouseEvent<any>) => {
    e.preventDefault();
    e.stopPropagation()
    return false;
  }

  return (
    <>
      <div
        onContextMenu={ignoreContextDropdown}
        className={`${recording ? styles.animatedCircle : ''}`}>
      </div>
      <div
        onContextMenu={ignoreContextDropdown}
        className={styles.container}>

        <div
          ref={imageRef}
          className={styles.outerCircle}
          onContextMenu={ignoreContextDropdown}
        >
          <div
            className={styles.innerCircle}
            onContextMenu={ignoreContextDropdown}
          >
            <div
              className={styles.innerMostCircle}
              onContextMenu={ignoreContextDropdown}
            ></div>
          </div>
        </div>
      </div>
    </>
  );
}

export default LongPressButton;
