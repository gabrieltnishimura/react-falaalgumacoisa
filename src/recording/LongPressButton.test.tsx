import { render } from '@testing-library/react';
import React from 'react';
import LongPressButton from './LongPressButton';

test('renders microphone button', () => {
  const start = jest.fn();
  const stop = jest.fn();
  const element = render(<LongPressButton pressed={start} unpressed={stop} />);

  const { getByText, container } = element;
  const linkElement = getByText(/Pressione o microfone para falar/i);

  expect(linkElement).not.toBeUndefined();
  expect(container).toMatchSnapshot()
});