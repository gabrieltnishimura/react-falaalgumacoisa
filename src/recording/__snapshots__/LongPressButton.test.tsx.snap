// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders microphone button 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="outerCircle"
    >
      <div
        class="innerCircle"
      >
        <div
          class="innerMostCircle"
        />
      </div>
    </div>
  </div>
</div>
`;
